### PR TITLE
Implements getopt for avr

### LINKED ISSUE #3355: Provide getopt for platforms that don't support it
newlib provides `getopt.h` which would implementing commands like `ping6` far easier. Sadly msp430-libc and avr-libc do not provide this header. Conclusively, a lightweight implementation for those platforms would be much appreciated.

### PR DESCRIPTION
**Pull request by virajsahai**
### Contribution description
Implemented getopt for atmega

### Issues/PRs references
Fixes #3355

### CODE REVIEW

File: .gitattributes [hunk 1]
```diff
@@ -5,3 +5,5 @@
 # when the heading is exactly 7 characters long.
 *.md conflict-marker-size=100
 *.txt conflict-marker-size=100
+*.c text eof=lf
```
Comment (aabadie) [inline .gitattributes]: This looks unrelated. Please provide those files in a separate PR if really required on your side.

File: cpu/atmega_common/getopt.c [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+char* optarg;
+int optind = 1, opterr = 1, optopt;
+
+int getopt(int argc, char* const argv[], const char* optstring)
+{
+    char* loc_optstr = NULL, loc_argvptr = NULL;
+    char opt;
+    optarg = NULL;
+
+    if(optind < 1 || optind > argc)
```
Comment (aabadie) [inline cpu/atmega_common/getopt.c]: Lot's of style issues, you can run uncrustify to format your files with a RIOT compliant style.

File: cpu/atmega_common/include/getopt.h [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+*/
+
+#ifndef GETOPT_H
+#define GETOPT_H
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+typedef struct{
```
Comment (aabadie) [inline cpu/atmega_common/include/getopt.h]: This struct is not documented.

File: cpu/atmega_common/include/getopt.h [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+    char* optarg;
+    int optind;
+    int opterr;
+    int optopt;
+} opt_t;
+
+static const opt_t def_opt = {NULL,1,1,-1};
```
Comment (aabadie) [inline cpu/atmega_common/include/getopt.h]: not documented

File: cpu/atmega_common/include/getopt.h [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+extern int optind, opterr, optopt;
+
+/**
+* @brief      Non re-entrant version of getopt.
+*
+* @param[in]  argc           The count of the number of arguments.
+* @param[in]  argv           The array containing all the arguments.
+* @param[in]  optstring      The string of expected options.
+*
+* @return ASCII value of the option if option was succesfully found, '?' or ':' when unknown option is found or argument is missing.
```
Comment (aabadie) [inline cpu/atmega_common/include/getopt.h]: This line is too long and should be split

File: doc/doxygen/riot.doxyfile [hunk 6]
```diff
@@ -826,6 +826,7 @@ EXCLUDE_PATTERNS       = */board/*/tools/* \
                          */cpu/msp430_common/include/sys/*.h \
                          */cpu/native/osx-libc-extra \
                          */cpu/x86/include/* \
+                         */cpu/atmega_common/include/getopt.h \
```
Comment (aabadie) [inline doc/doxygen/riot.doxyfile]: why ?

**Comment by miri64** [comment]
Also, please provide a more telling commit message for your commit. It should tell what module got change and what the change entails. Our usual format would be something like

```
atmega_common: provide getopt support
```

File: .gitattributes (hunk 1 above)
Comment (virajsahai) [reply .gitattributes]: I had to do this to get LF after every line. Please do let me know if there is any other way of getting rid of them. Thanks!

File: cpu/atmega_common/getopt.c (hunk 2 above)
Comment (virajsahai) [reply cpu/atmega_common/getopt.c]: Will do! Thanks

File: cpu/atmega_common/include/getopt.h (hunk 3 above)
Comment (virajsahai) [reply cpu/atmega_common/include/getopt.h]: I am sorry for being extra lost but I am fairly new to Open Source. I went through all the dev recommendations but didn't find the location where to document it?

File: cpu/atmega_common/include/getopt.h (hunk 5 above)
Comment (virajsahai) [reply cpu/atmega_common/include/getopt.h]: Will do. Thanks!

**Comment by virajsahai** [comment]
@aabadie Thanks for all the helpful comments. I have fixed most issues. just need to document the struct. Please let me know where and how, couldn't find the relevant doc. Thanks!!

File: cpu/atmega_common/include/getopt.h [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+ */
+
+#ifndef GETOPT_H
+#define GETOPT_H
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+typedef struct {
```
Comment (aabadie) [inline cpu/atmega_common/include/getopt.h]: To document this, just add a doxygen `@brief` before the struct. Each attribute must be document as well. See [here](https://github.com/RIOT-OS/RIOT/blob/master/core/include/msg.h#L176) for an example.

File: cpu/atmega_common/getopt.c [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+                    if (optstring[0] == ':') {
+                        return ":";
+                    }
+                    return "?";
+                }
+            }
+        }
+    }
+    return c;
+}
```
Comment (aabadie) [inline cpu/atmega_common/getopt.c]: there's a missing newline at the end of this file.

File: cpu/atmega_common/include/getopt.h [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+extern int optind, opterr, optopt;
+
+/**
+ * @brief      Non re-entrant version of getopt.
+ *
+ * @param[in]  argc           The count of the number of arguments.
+ * @param[in]  argv           The array containing all the arguments.
+ * @param[in]  optstring      The string of expected options.
+ *
+ * @return ASCII value of the option if option was succesfully found.
```
Comment (aabadie) [inline cpu/atmega_common/include/getopt.h]: typo on succes**s**fully

File: cpu/atmega_common/include/getopt.h [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+
+/**
+ * @brief      Re-entrant version of getopt.
+ *
+ * @param[in]  argc           The count of the number of arguments.
+ * @param[in]  argv           The array containing all the arguments.
+ * @param[in]  optstring      The string of expected options.
+ * @param[out] r              The per call struct for holding return args and other options.
+ *
+ * @return ASCII value of the option if option was succesfully found.
```
Comment (aabadie) [inline cpu/atmega_common/include/getopt.h]: typo on succes**s**fully

File: cpu/atmega_common/include/getopt.h [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+ *         '?' or ':' when unknown option is found or argument is missing.
+ */
+
```
Comment (aabadie) [inline cpu/atmega_common/include/getopt.h]: this blank line can be removed

File: cpu/atmega_common/include/getopt.h [hunk 12]
```diff
@@ -0,0 +1,77 @@
...
+ * @brief      Non re-entrant version of getopt.
+ *
+ * @param[in]  argc           The count of the number of arguments.
+ * @param[in]  argv           The array containing all the arguments.
+ * @param[in]  optstring      The string of expected options.
+ *
+ * @return ASCII value of the option if option was succesfully found.
+ *         '?' or ':' when unknown option is found or argument is missing.
+ */
+
```
Comment (aabadie) [inline cpu/atmega_common/include/getopt.h]: this blank line can be removed

**Comment by virajsahai** [comment]
@aabadie i have done some generic testing with some edge cases but have done the real deal. It'd be great if someone can advise on that. Thanks!

**Comment by virajsahai** [comment]
@miri64 @aabadie any updates??

File: cpu/atmega_common/getopt.c [hunk 13]
```diff
@@ -0,0 +1,155 @@
...
+ */
+
+/**
+ * @ingroup     cpu_atmega_common
+ * @{
+ *
+ * @file        getopt.c
+ * @brief       Implementation of getopt lib.
+ *
+ * This implements the getopt and getopt_r functionality. getopt_r has a few caveats, make sure to use it correctly.
```
Comment (aabadie) [inline cpu/atmega_common/getopt.c]: This line should be split

File: cpu/atmega_common/getopt.c [hunk 14]
```diff
@@ -0,0 +1,155 @@
...
+    char *loc_argvptr = NULL;
+    char opt;
+
+    optarg = NULL;
+
+    if (optind < 1 || optind >= argc) {
+        return -1;
+    }
+
+    if (*argv[optind] != '-' || !(argv[optind] + 1) || *(argv[optind] + 1) == '\0' || *(argv[optind] + 1) == '-') {
```
Comment (aabadie) [inline cpu/atmega_common/getopt.c]: This line should be split

File: cpu/atmega_common/getopt.c [hunk 15]
```diff
@@ -0,0 +1,155 @@
...
+{
+    char *loc_optstr = NULL;
+    char *loc_argvptr = NULL;
+    char opt;
+
+    if (r->optind < 1 || r->optind >= argc) {
+        return -1;
+    }
+
+    if (*argv[r->optind] != '-' || !(argv[r->optind] + 1) || *(argv[r->optind] + 1) == '\0' || *(argv[r->optind] + 1) == '-') {
```
Comment (aabadie) [inline cpu/atmega_common/getopt.c]: This line should be split

File: cpu/atmega_common/getopt.c (hunk 13 above)
Comment (aabadie) [reply cpu/atmega_common/getopt.c]: In fact, it should be removed, it's a copy-paste of the one in the header file.

File: cpu/atmega_common/include/getopt.h [hunk 16]
```diff
@@ -0,0 +1,83 @@
...
+/**
+ * @ingroup     cpu_atmega_common
+ * @{
+ *
+ * @file        header
+ * @brief       Implementation of getopt lib.
+ *
+ * This implements the getopt and getopt_r functionality. getopt_r has a few caveats, make sure to use it correctly.
```
Comment (aabadie) [inline cpu/atmega_common/include/getopt.h]: This line should be split

File: cpu/atmega_common/include/getopt.h [hunk 17]
```diff
@@ -0,0 +1,83 @@
...
+    int optind;
+    int opterr;
+    int optopt;
+} opt_t;
+
+/*Don't forget to use this else the call to getopt_r will fail eg. opt_t *x = INIT_OPT_T_PTR;*/
+/*INIT_OPT_T_PTR is used to initalize an opt_t* variable, INIT_OPT_T for opt_t. */
+/*To change options, edit the fields of opt_t and call getopt_r*/
+
+#define INIT_OPT_T_PTR &(opt_t){ NULL, 1, 1, -1 }
```
Comment (aabadie) [inline cpu/atmega_common/include/getopt.h]: Those defines should be documented

**Comment by miri64** [comment]
> @aabadie i have done some generic testing with some edge cases but haven't done the real deal. It'd be great if someone can advise on that. Thanks!

Not sure why I specifically was summoned for this, but if you have test cases, please provide them as an application in `tests/`.

**Comment by aabadie** [comment]
> Not sure why I specifically was summoned for this

because you opened #3355 ;)

**Comment by virajsahai** [comment]
@aabadie by generic testing, I meant testing the logic completeness in edge cases. I don't know how to test after integrating.

**Comment by virajsahai** [comment]
@aabadie any updates??

**Comment by aabadie** [comment]
> I meant testing the logic completeness in edge cases

I don't understand this, sorry. What edge cases ? How do you do that ? Do you have a RIOT specific application for that ?

**Comment by virajsahai** [comment]
@aabadie No..i don't have a riot specific application for that. It would be great if you could point me to someone who can help with the testing.

And I tested the code as a regular c code on my pc..for cases when double colons are there..empty string passed etc.

**Comment by virajsahai** [comment]
@aabadie any updates?

**Comment by miri64** [comment]
@jcarrano this somewhat coincides with your work in #9538. Can you have a look as well?

**Comment by jcarrano** [comment]
#9538 and this are not quite the same (#9538 does way more than getopt).

The question here is whether we seek to suport POSIX (or at least a reasonable subset). If we do, then regardless of #9538 this should go in.
